File: packages/ios/Sources/UnforgettableSDK/UnforgettableSDK.swift
import Foundation

/// The mode for the Unforgettable SDK
public enum UnforgettableMode: String, Codable {
    case create
    case restore
}

/// Options for initializing the Unforgettable SDK
public struct UnforgettableSdkOptions {
    public let mode: UnforgettableMode
    public let appUrl: String
    public let apiUrl: String
    public let factors: [RecoveryFactor]
    public let walletAddress: String?
    public let group: String?
    public let customParams: [String: String]?
    
    public init(
        mode: UnforgettableMode,
        appUrl: String = UNFORGETTABLE_APP_URL,
        apiUrl: String = UNFORGETTABLE_API_URL,
        factors: [RecoveryFactor] = [],
        walletAddress: String? = nil,
        group: String? = nil,
        customParams: [String: String]? = nil
    ) {
        self.mode = mode
        self.appUrl = appUrl
        self.apiUrl = apiUrl
        self.factors = factors
        self.walletAddress = walletAddress
        self.group = group
        self.customParams = customParams
    }
}

/// Payload inside the data transfer
struct DataTransferPayload: Codable {
    let recovery_key: String
    let helper_data_url: String?
}

/// Recovered data from the API
public struct RecoveredData {
    public let recoveryKey: String
    public let helperDataUrl: String?
    
    public init(recoveryKey: String, helperDataUrl: String? = nil) {
        self.recoveryKey = recoveryKey
        self.helperDataUrl = helperDataUrl
    }
}

/// Errors that can occur in the SDK
public enum UnforgettableSDKError: Error {
    case networkError(Error)
    case invalidResponse
    case notFound
    case decodingError(Error)
    case cryptoError(Error)
}

/// Main SDK class for Unforgettable
public class UnforgettableSDK {
    public let mode: UnforgettableMode
    public let appUrl: String
    public let factors: [RecoveryFactor]
    public let walletAddress: String?
    public let group: String?
    public let customParams: [String: String]?
    
    private let apiUrl: String
    private let dataTransferId: String
    private let encryptionKeyPair: DataTransferKeyPair
    
    /// Initializes the Unforgettable SDK
    /// - Parameter options: Configuration options
    /// - Throws: CryptoError if key generation fails
    public init(options: UnforgettableSdkOptions) throws {
        self.mode = options.mode
        self.appUrl = options.appUrl
        self.apiUrl = options.apiUrl
        self.factors = options.factors
        self.walletAddress = options.walletAddress
        self.group = options.group
        self.customParams = options.customParams
        
        self.dataTransferId = UUID().uuidString.lowercased()
        self.encryptionKeyPair = try generateDataTransferKeyPair()
    }
    
    /// Generates the recovery URL
    /// - Returns: The recovery URL as a string
    public func getRecoveryUrl() -> String {
        let baseUrl = appUrl.hasSuffix("/") ? String(appUrl.dropLast()) : appUrl
        let path = mode == .restore ? "/r" : "/c"
        
        let urlString = baseUrl + path
        
        let hash = composeUnforgettableLocationHash(params: UnforgettablePathParams(
            dataTransferId: dataTransferId,
            encryptionPublicKey: encryptionKeyPair.publicKey,
            factors: factors,
            walletAddress: walletAddress,
            group: group,
            customParams: customParams
        ))
        
        return urlString + hash
    }
    
    /// Retrieves the recovered data from the API
    /// - Returns: The recovered data
    /// - Throws: UnforgettableSDKError on failure
    public func getRecoveredData() async throws -> RecoveredData {
        let endpoint = "/integrations/helper-keeper/v1/public/data-transfers/\(dataTransferId)"
        let urlString = apiUrl + endpoint
        
        guard let url = URL(string: urlString) else {
            throw UnforgettableSDKError.invalidResponse
        }
        
        let (data, response) = try await URLSession.shared.data(from: url)
        
        guard let httpResponse = response as? HTTPURLResponse else {
            throw UnforgettableSDKError.invalidResponse
        }
        
        guard httpResponse.statusCode == 200 else {
            if httpResponse.statusCode == 404 {
                throw UnforgettableSDKError.notFound
            }
            throw UnforgettableSDKError.invalidResponse
        }
        
        // Parse the response
        let decoder = JSONDecoder()
        let dataTransferResponse: DataTransferResponse
        
        do {
            dataTransferResponse = try decoder.decode(DataTransferResponse.self, from: data)
        } catch {
            throw UnforgettableSDKError.decodingError(error)
        }
        
        guard let dataTransferWrapper = dataTransferResponse.data else {
            throw UnforgettableSDKError.notFound
        }
        
        // Parse the nested payload from attributes.data
        let payload: DataTransferPayload
        do {
            guard let payloadData = dataTransferWrapper.attributes.data.data(using: .utf8) else {
                throw UnforgettableSDKError.decodingError(NSError(domain: "UnforgettableSDK", code: -1))
            }
            payload = try decoder.decode(DataTransferPayload.self, from: payloadData)
        } catch {
            throw UnforgettableSDKError.decodingError(error)
        }
        
        // The recovery_key contains raw binary data incorrectly encoded as a UTF-8 string
        // We need to convert it back to Data by reading the raw bytes
        guard let encryptedData = payload.recovery_key.data(using: .isoLatin1) else {
            throw UnforgettableSDKError.decodingError(NSError(domain: "UnforgettableSDK", code: -2))
        }
        
        // Decrypt the recovery key using the binary data directly
        let decryptedKey: String
        do {
            decryptedKey = try encryptionKeyPair.decryptBinary(encryptedData)
        } catch {
            throw UnforgettableSDKError.cryptoError(error)
        }
        
        return RecoveredData(
            recoveryKey: decryptedKey,
            helperDataUrl: payload.helper_data_url
        )
    }
    
    /// Retrieves only the recovered key
    /// - Returns: The recovered key as a string
    /// - Throws: UnforgettableSDKError on failure
    public func getRecoveredKey() async throws -> String {
        let recoveredData = try await getRecoveredData()
        return recoveredData.recoveryKey
    }
}

// MARK: - Response Models

private struct DataTransferResponse: Codable {
    let data: DataTransferWrapper?
}

private struct DataTransferWrapper: Codable {
    let id: String
    let type: String
    let attributes: DataTransferAttributes
}

private struct DataTransferAttributes: Codable {
    let data: String
}

File: packages/android/src/main/kotlin/com/rarimo/unforgettable/UnforgettableSDK.kt
package com.rarimo.unforgettable

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.SerialName
import kotlinx.serialization.json.Json
import java.net.HttpURLConnection
import java.net.URL
import java.util.UUID

/**
 * The mode for the Unforgettable SDK
 */
enum class UnforgettableMode {
    CREATE,
    RESTORE
}

/**
 * Options for initializing the Unforgettable SDK
 */
data class UnforgettableSdkOptions(
    val mode: UnforgettableMode,
    val appUrl: String = UNFORGETTABLE_APP_URL,
    val apiUrl: String = UNFORGETTABLE_API_URL,
    val factors: List<RecoveryFactor> = emptyList(),
    val walletAddress: String? = null,
    val group: String? = null,
    val customParams: Map<String, String>? = null
)

/**
 * API response wrapper
 */
@Serializable
private data class DataTransferResponse(
    val data: DataTransferWrapper? = null
)

@Serializable
private data class DataTransferWrapper(
    val id: String,
    val type: String,
    val attributes: DataTransferAttributes
)

@Serializable
private data class DataTransferAttributes(
    val data: String
)

/**
 * Payload inside the data transfer
 */
@Serializable
private data class DataTransferPayload(
    @SerialName("recovery_key")
    val recoveryKey: String,
    @SerialName("helper_data_url")
    val helperDataUrl: String? = null
)

/**
 * Recovered data from the API
 */
data class RecoveredData(
    val recoveryKey: String,
    val helperDataUrl: String? = null
)

/**
 * Errors that can occur in the SDK
 */
sealed class UnforgettableSDKError : Exception() {
    data class NetworkError(override val cause: Throwable) : UnforgettableSDKError()
    object InvalidResponse : UnforgettableSDKError()
    object NotFound : UnforgettableSDKError()
    data class DecodingError(override val cause: Throwable) : UnforgettableSDKError()
    data class CryptoError(override val cause: Throwable) : UnforgettableSDKError()
}

/**
 * Main SDK class for Unforgettable
 */
public class UnforgettableSDK(options: UnforgettableSdkOptions) {
    val mode: UnforgettableMode = options.mode
    val appUrl: String = options.appUrl
    val factors: List<RecoveryFactor> = options.factors
    val walletAddress: String? = options.walletAddress
    val group: String? = options.group
    val customParams: Map<String, String>? = options.customParams
    
    private val apiUrl: String = options.apiUrl
    private val dataTransferId: String = UUID.randomUUID().toString().lowercase()
    private val encryptionKeyPair: DataTransferKeyPair = generateDataTransferKeyPair()
    
    private val json = Json {
        ignoreUnknownKeys = true
        isLenient = true
    }
    
    /**
     * Generates the recovery URL
     * @return The recovery URL as a string
     */
    fun getRecoveryUrl(): String {
        val baseUrl = if (appUrl.endsWith("/")) appUrl.dropLast(1) else appUrl
        val path = if (mode == UnforgettableMode.RESTORE) "/r" else "/c"
        
        val hash = composeUnforgettableLocationHash(
            UnforgettablePathParams(
                dataTransferId = dataTransferId,
                encryptionPublicKey = encryptionKeyPair.publicKey,
                factors = factors,
                walletAddress = walletAddress,
                group = group,
                customParams = customParams
            )
        )
        
        return "$baseUrl$path$hash"
    }
    
    /**
     * Retrieves the recovered data from the API
     * @return The recovered data
     * @throws UnforgettableSDKError on failure
     */
    suspend fun getRecoveredData(): RecoveredData = withContext(Dispatchers.IO) {
        val endpoint = "/integrations/helper-keeper/v1/public/data-transfers/$dataTransferId"
        val urlString = "$apiUrl$endpoint"
        
        try {
            val url = URL(urlString)
            val connection = url.openConnection() as HttpURLConnection
            
            try {
                connection.requestMethod = "GET"
                connection.connectTimeout = 30000
                connection.readTimeout = 30000
                
                when (connection.responseCode) {
                    HttpURLConnection.HTTP_OK -> {
                        val responseBody = connection.inputStream.bufferedReader().use { it.readText() }
                        
                        val response = try {
                            json.decodeFromString<DataTransferResponse>(responseBody)
                        } catch (e: Exception) {
                            throw UnforgettableSDKError.DecodingError(e)
                        }
                        
                        val dataTransferWrapper = response.data
                            ?: throw UnforgettableSDKError.NotFound
                        
                        val payload = try {
                            json.decodeFromString<DataTransferPayload>(dataTransferWrapper.attributes.data)
                        } catch (e: Exception) {
                            throw UnforgettableSDKError.DecodingError(e)
                        }
                        
                        // The recovery_key contains raw binary data incorrectly encoded as a UTF-8 string
                        // We need to convert it back to bytes using ISO-8859-1 encoding
                        val encryptedBytes = payload.recoveryKey.toByteArray(Charsets.ISO_8859_1)
                        
                        val decryptedKey = try {
                            encryptionKeyPair.decryptBinary(encryptedBytes)
                        } catch (e: Exception) {
                            throw UnforgettableSDKError.CryptoError(e)
                        }
                        
                        RecoveredData(
                            recoveryKey = decryptedKey,
                            helperDataUrl = payload.helperDataUrl
                        )
                    }
                    HttpURLConnection.HTTP_NOT_FOUND -> {
                        throw UnforgettableSDKError.NotFound
                    }
                    else -> {
                        throw UnforgettableSDKError.InvalidResponse
                    }
                }
            } finally {
                connection.disconnect()
            }
        } catch (e: UnforgettableSDKError) {
            throw e
        } catch (e: Exception) {
            throw UnforgettableSDKError.NetworkError(e)
        }
    }
    
    /**
     * Retrieves only the recovered key
     * @return The recovered key as a string
     * @throws UnforgettableSDKError on failure
     */
    suspend fun getRecoveredKey(): String {
        val recoveredData = getRecoveredData()
        return recoveredData.recoveryKey
    }
}
